intdataplane/int_dataplane.go

```diff
@@ -750,3 +752,39 @@ type iptablesTable interface {
 	RemoveChains([]*iptables.Chain)
 	RemoveChainByName(name string)
 }
+
+// msgStringer wraps an API message to customise how we stringify it.  For example, it truncates
+// the lists of members in the (potentially very large) IPSetsUpdate messages.
+type msgStringer struct {
+	msg interface{}
+}
+
+func (m msgStringer) String() string {
+	if log.GetLevel() < log.DebugLevel && m.msg != nil {
+		const truncateAt = 10
+		switch msg := m.msg.(type) {
+		case *proto.IPSetUpdate:
+			if len(msg.Members) < truncateAt {
+				return fmt.Sprintf("%v", msg)
+			}
+			return fmt.Sprintf("id:%#v members(%d):%#v(truncated)",
+				msg.Id, len(msg.Members), msg.Members[:truncateAt])
+		case *proto.IPSetDeltaUpdate:
+			if len(msg.AddedMembers) < truncateAt && len(msg.RemovedMembers) < truncateAt {
+				return fmt.Sprintf("%v", msg)
+			}
+			addedNum := truncateAt
+			removedNum := truncateAt
+			if len(msg.AddedMembers) < addedNum {
+				addedNum = len(msg.AddedMembers)
+			}
+			if len(msg.RemovedMembers) < removedNum {
+				removedNum = len(msg.RemovedMembers)
+			}
```

Comment:
If these calculations of the `X` in `<slice>[:X]` are needed here, why aren't they also needed in the IPSetUpdate case just above?

(If they are needed, I guess that's because `a[:X]` will panic if `len(a) < X`.  Is that correct?)